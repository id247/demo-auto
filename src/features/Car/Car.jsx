import React from 'react';
import PropTypes from 'prop-types';
import { carPropType, fetchStatesPropType } from 'constants/propTypes';
import FetchError from 'components/FetchError';
import CarLayout from 'components/CarLayout';
import FavoriteCar from 'components/FavoriteCar';
import CarDescription from 'components/CarDescription';

class Car extends React.Component {
  componentDidMount() {
    this.fetchData();
  }

  fetchData = () => {
    const { fetchCarAsync, carId } = this.props;
    fetchCarAsync({ carId });
  };

  render() {
    const {
      car = {},
      carId,
      fetchStates,
      toggleFavoriteCar,
      isFavorite
    } = this.props;

    if (fetchStates.isFetchingError) {
      return <FetchError fetchData={this.fetchData} />;
    }

    return (
      <CarLayout
        carImage={car.pictureUrl}
        carTitle={car.title}
        fetchStates={fetchStates}
        info={<CarDescription car={car} fetchStates={fetchStates} />}
        sidebar={
          <FavoriteCar
            carId={carId}
            isFavorite={isFavorite}
            fetchStates={fetchStates}
            toggleFavoriteCar={toggleFavoriteCar}
          />
        }
      />
    );
  }
}

Car.propTypes = {
  car: carPropType,
  fetchStates: fetchStatesPropType,
  carId: PropTypes.number.isRequired,
  fetchCarAsync: PropTypes.func.isRequired,
  toggleFavoriteCar: PropTypes.func.isRequired,
  isFavorite: PropTypes.bool.isRequired
};

export default Car;
