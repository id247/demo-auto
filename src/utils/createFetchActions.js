export default (prefix, name) => {
  const requestType = `${prefix}${name}_REQUEST`;
  const successType = `${prefix}${name}_SUCCESS`;
  const failType = `${prefix}${name}_FAIL`;

  return {
    requestType,
    successType,
    failType,
    fetch(payload) {
      return {
        type: requestType,
        payload
      };
    },
    success(payload) {
      return {
        type: successType,
        payload
      };
    },
    fail(payload) {
      return {
        type: failType,
        payload
      };
    }
  };
};
