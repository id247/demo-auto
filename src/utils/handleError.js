import history from 'myHistory';

export const handleError = e => {
  if (!e) return;

  if (e.response && e.response.status === 404) {
    history.replace('/404');
  } else {
    console.error(e);
  }
};

export default handleError;
