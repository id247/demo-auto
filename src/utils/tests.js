import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import createStore from 'store';

export const testActionCreator = ({ type, fn }) => {
  describe(fn.name, () => {
    it('should return correct action', () => {
      expect(fn({})).toEqual({ type, payload: {} });
    });

    it('should return undefined payload if non was passed', () => {
      expect(fn()).toEqual({ type, payload: undefined });
    });
  });
};

export const renderWithReduxAndRouter = (
  ui,
  {
    initialState,
    store = createStore(initialState),
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <Router history={history}>{ui}</Router>
    </Provider>
  ),
  store,
  history
});
