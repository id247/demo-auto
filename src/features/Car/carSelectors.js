import { createSelector } from 'reselect';
import { getFetchStates } from 'utils/stateMachine';

const carSelector = state => state.car;

export const carFetchStatesSelector = createSelector(
  carSelector,
  car => getFetchStates(car.fetchState)
);
