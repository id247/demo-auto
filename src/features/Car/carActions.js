import apiRequest from 'api';
import apiRoutes from 'api/api-routes';
import { adaptCar } from 'adapters/cars';
import { addCar } from 'actions/cars';
import createFetchActions from 'utils/createFetchActions';
import handleError from 'utils/handleError';

const prefix = 'car/';

export const fetchCarActions = createFetchActions(prefix, 'FETCH_CAR');

export const fetchCarAsync = ({ carId } = {}) => async dispatch => {
  if (!carId) {
    return;
  }

  dispatch(fetchCarActions.fetch());

  try {
    const car = await apiRequest(`${apiRoutes.cars}/${carId}`);
    dispatch(addCar({ car: adaptCar(car) }));
    dispatch(fetchCarActions.success());
  } catch (e) {
    dispatch(fetchCarActions.fail());
    handleError(e);
  }
};
