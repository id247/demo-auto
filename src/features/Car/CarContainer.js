import { connect } from 'react-redux';
import { fetchCarAsync } from './carActions';
import { carByIdSelector, isFavoriteSelector } from 'selectors/cars';
import { carFetchStatesSelector } from './carSelectors';
import { toggleFavoriteCar } from 'actions/favorites';
import Car from './Car';

export default connect(
  (state, ownProps) => ({
    car: carByIdSelector(state, ownProps),
    fetchStates: carFetchStatesSelector(state),
    isFavorite: isFavoriteSelector(state, ownProps)
  }),
  {
    fetchCarAsync,
    toggleFavoriteCar
  }
)(Car);
