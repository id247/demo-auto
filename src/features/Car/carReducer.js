import fetchStates from 'constants/fetchStates';
import { getNewStateMachineState } from 'utils/stateMachine';
import { fetchCarActions } from './carActions';

const stateMachine = {
  [fetchStates.idle]: {
    name: fetchStates.idle,
    to: {
      [fetchCarActions.successType]: fetchStates.done,
      [fetchCarActions.failType]: fetchStates.error
    }
  },
  [fetchStates.fetching]: {
    name: fetchStates.fetching,
    to: {
      [fetchCarActions.successType]: fetchStates.done,
      [fetchCarActions.failType]: fetchStates.error
    }
  },
  [fetchStates.done]: {
    name: fetchStates.done,
    to: {
      [fetchCarActions.requestType]: fetchStates.fetching
    }
  },
  [fetchStates.error]: {
    name: fetchStates.error,
    to: {
      [fetchCarActions.requestType]: fetchStates.idle
    }
  }
};

const initialState = {
  fetchState: fetchStates.idle
};

const carSearchReducer = (state = initialState, { type, payload }) => {
  const fetchState = getNewStateMachineState(
    stateMachine,
    state.fetchState,
    type
  );

  switch (type) {
    case fetchCarActions.requestType:
    case fetchCarActions.successType:
    case fetchCarActions.failType:
      return {
        ...state,
        fetchState
      };
    default:
      return state;
  }
};

export default carSearchReducer;
