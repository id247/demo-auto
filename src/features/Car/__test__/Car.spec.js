import 'jest-dom/extend-expect';
import React from 'react';
import { cleanup, wait } from 'react-testing-library';
import axios from 'axios';
import { renderWithReduxAndRouter } from 'utils/tests';
import Car from '../';

afterEach(() => {
  // eslint-disable-next-line no-console
  console.error = jest.fn();
  cleanup();
  axios.mockReset();
});

describe('Car component', () => {
  const carBE = {
    color: 'red',
    fuelType: 'Diesel',
    manufacturerName: 'Mercedes-Benz',
    mileage: { number: 158099, unit: 'km' },
    unit: 'km',
    modelName: 'A-Klasse',
    pictureUrl: 'http://localhost/car.jpg',
    id: 10252,
    text: 'description'
  };

  const props = {
    carId: 10252,
    fetchStates: {
      isIdle: true,
      isFetching: false,
      isFetchingDone: false,
      isFetchingError: false
    }
  };

  test('Initial idle render', async () => {
    axios.mockImplementation(o => {
      return Promise.resolve({ data: carBE });
    });

    const { container } = renderWithReduxAndRouter(<Car {...props} />);

    expect(container).toMatchSnapshot();
  });

  test('Render after successful fetching ', async () => {
    axios.mockImplementation(o => {
      return Promise.resolve({ data: carBE });
    });

    const { container, getByText } = renderWithReduxAndRouter(
      <Car {...props} />
    );

    expect(axios).toHaveBeenCalledTimes(1);

    await wait(() => getByText('Mercedes-Benz A-Klasse'));

    expect(container).toMatchSnapshot();
  });

  test('Render after unsuccessful fetching', async () => {
    axios.mockImplementation(o => {
      return Promise.reject();
    });

    const { container, getByText } = renderWithReduxAndRouter(
      <Car {...props} />
    );

    expect(axios).toHaveBeenCalledTimes(1);

    await wait(() => getByText('Something went wrong'));

    expect(container).toMatchSnapshot();
  });
});
