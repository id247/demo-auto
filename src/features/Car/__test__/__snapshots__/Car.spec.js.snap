// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Car component Initial idle render 1`] = `
<div>
  <div
    class="car-layout"
  >
    <div
      class="car-layout__image-wrapper"
    />
    <div
      class="wrapper wrapper--m"
    >
      <div
        class="car-layout__content"
      >
        <div
          class="car-layout__info"
        >
          <div
            class="car-description"
          >
            <h1
              class="car-description__title"
            >
              <span
                class="idle"
                style="height: 30px;"
              >
                 
              </span>
            </h1>
            <div
              class="car-description__info"
            >
              <span
                class="idle"
                style="height: 20px;"
              >
                 
              </span>
            </div>
            <span
              class="idle"
              style="height: 60px;"
            >
               
            </span>
          </div>
        </div>
        <div
          class="car-layout__sidebar"
        >
          <div
            class="sidebar-item"
          >
            <div
              class="text"
            >
              <span
                class="idle"
                style="height: 40px;"
              >
                 
              </span>
            </div>
            <div
              class="sidebar-button"
            >
              <span
                class="idle"
                style="display: inline-block; width: 128px; height: 30px;"
              >
                 
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Car component Render after successful fetching  1`] = `
<div>
  <div
    class="car-layout"
  >
    <div
      class="car-layout__image-wrapper"
    >
      <img
        alt="Mercedes-Benz A-Klasse"
        class="car-layout__image"
        src="http://localhost/car.jpg"
      />
    </div>
    <div
      class="wrapper wrapper--m"
    >
      <div
        class="car-layout__content"
      >
        <div
          class="car-layout__info"
        >
          <div
            class="car-description"
          >
            <h1
              class="car-description__title"
            >
              Mercedes-Benz A-Klasse
            </h1>
            <div
              class="car-description__info"
            >
              Stock # 10252 - 158099 km - Diesel - red
            </div>
            <div
              class="car-description__text"
            >
              <div
                class="text"
              >
                <p>
                  description
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="car-layout__sidebar"
        >
          <div
            class="sidebar-item"
          >
            <div
              class="text"
            >
              <p>
                If you like this car, click the button and save it in your collection of favourite items.
              </p>
            </div>
            <div
              class="sidebar-button"
            >
              <button
                class="button button--primary"
                data-testid="button"
                type="button"
              >
                Save
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Car component Render after unsuccessful fetching 1`] = `
<div>
  <div
    class="error"
  >
    <div
      class="error__text"
    >
      Something went wrong
    </div>
    <button
      class="button button--primary"
      data-testid="button"
      type="button"
    >
      Retry
    </button>
  </div>
</div>
`;
